[adam07018] iP

adam07018 opened the pull request:
### **Duke**
`You may delay, but time will not. -Miles Davis`
### Always lost track of the deadline:scream:? Have a bad memory?:face_with_spiral_eyes:

Don't worry! Now you have Duke to help you. It can do following:
1. Add different kinds of task: Todo, Deadline, Event
4. Mark a task as done or not done yet
5. List out all the tasks you have written down
6. Delete any tasks from your list
7. Search the tasks you created by keywords

### Why is it so good?

- **Easy** to learn, there are examples provided for you:partying_face:
- Fast speed. **Run and Go**
- Record down your task lists into hardcopy
- **Free!!**
### How to use? Easy lah. Even my ah mah can do it!
- [ ] download it from [here](https://github.com/adam07018/ip).
- [ ] double-click it.
- [ ] add your tasks.
- [ ] And here we go!
### **Time doesn't wait!:star_struck: Download and use now to make your life much easier!**

adam07018 committed:
no message

adam07018 committed:
level1

adam07018 committed:
lv2

adam07018 committed:
lv3

adam07018 committed:
lv4

adam07018 committed:
level5

adam07018 committed:
level6

adam07018 committed:
implement level_6

adam07018 committed:
Updates file writing system

adam07018 committed:
add Time recognition

adam07018 committed:
add graddle from remote repo

adam07018 committed:
Gradle for checkstyle

adam07018 committed:
Junit testcase

adam07018 committed:
Enum for exception

adam07018 committed:
More oop version 1

adam07018 committed:
absolute OOP

adam07018 committed:
pacakge

adam07018 committed:
nothing

adam07018 committed:
nothing

adam07018 committed:
Add javadoc
Catch exception when time format is wrong
Fix bugs that app will stall when exception is met

adam07018 committed:
Fix checkstyle for switch

adam07018 committed:
Add find method

adam07018 committed:
Merge branch 'branch-A-CodingStandard'
Solve conflict

adam07018 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java

adam07018 committed:
Adds all javadoc for classes
Fixes checkstyle

huanghao1998 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class DukeException extends Exception {
+    DukeException(ExceptionType e) {
+        switch (e) {
```
I like you have used switch case to tackle different exception, but i would recommend to extend the different exceptions out into different exception class. That would be easier to print out exception message when implementing GUI.

ZHTang29 replied:
Likewise, I feel the exception handling could be further split into individual classes, but most exceptions are handled so well done on that.

huanghao1998 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Class that display messages (except error messages) to user
+ */
+public class Ui {
```
The naming of the method is well done

huanghao1998 commented on src/main/java/duke/Times.java:
```diff
@@ -0,0 +1,30 @@
+ * Input of time will be formatted to MMM d yyyy hh:mm
+ */
+public class Times {
+    static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
```
If ``FORMATTER`` is not used in other class and try to encapsulate it using ``private``

adam07018 replied:
Noted with thanks

huanghao1998 reviewed (commented):
Overall, the code is neat, naming of method and class are well defined :)

adam07018 committed:
Add gui
Restructure DukeException by creating children Exception class and toString method
Change return type of command method execute from void to String

ZHTang29 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+                continue;
+            } catch (NullPointerException e) {
+                continue;
+            }
```
I like the use of the parser, and overall the OOP is done up appropriately! Clear java-docs as well so well done on the documentation portion.

ZHTang29 reviewed (commented):
Overall, I find the code easy to read, because you follow proper naming conventions and the javadocs in generally are done up well. Other than the minor portion on splitting the exception classes (if necessary to you), well done and keep up the good work!

adam07018 committed:
Organize all classes into various packages

Unify all constructor to be public so that methods in other packages could access

Tidy up code so that it follows checkstyle

adam07018 committed:
Add assertion in a few places to ensure validity of arguments and indexes.

adam07018 committed:
Merge pull request #1 from adam07018/branch-A-Assertions

Add assertion in a few places to ensure validity of arguments and ind…

adam07018 committed:
Refract code quality

Add some final int variable to get rid of magic number

adam07018 committed:
Merge pull request #2 from adam07018/branch-A-CodeQuality

Refract code quality

adam07018 committed:
Add feature to detect duplicates

adam07018 committed:
Add Ui example in doc

adam07018 committed:
Add user guide
